evaluations/ipfs.md

```diff
@@ -0,0 +1,41 @@
+An evaluation of the IPFS paper
+===============================
+
+2021-01-06 Mark Spanbroek
+
+https://ipfs.io/ipfs/QmR7GSQM93Cx5eAg6a6yRzNde1FQv7uL6X1o4k7zrJa3LX/ipfs.draft3.pdf
+
+Goal of this evaluation is to find things to adopt or avoid while designing
+Dagger. It is not meant to be a criticism of IPFS.
+
+#### Pros:
+
++ IPFS is designed by simplifying, evolving, and connecting proven techniques
+  (§3)
++ Consists of a stack of separately described sub-protocols (§3)
++ Uses Coral DSHT to favor data that is nearby, reducing latency of lookup
+  (§2.1.2)
++ Uses proof-of-work in S/Kademlia to discourage Sybil attacks (§2.1.3)
++ Favors self-describing values such as multihash (§3.1) and multiaddr (§3.2.1)
++ BitSwap protocol for exchanging blocks supports multiple strategies (§3.4.2)
++ Uses content addressing (§3.5)
++ The Merkle DAG is simple, yet allows constucting filesystems,
+  key-value stores, databases, messaging system, etc.. (§3.5)
+
+#### Cons:
+
+- Kademlia prefers long-lived nodes, which is not ideal for mobile enviroments
+  (§2.1.1)
+- BitSwap falls just short of introducing a currency with micro payments,
+  necessitating additional work for nodes to find blocks to barter with (§3.4)
+- Merkle DAGs and Paths (§3.5.1) feel like an unnecessary complication at the
+  protocol level; they could have been implemented on top of a system consisting
+  solely of content addressable chunks
+- Object pinning (§3.5.3) inevitably leads to centralized gateways to IPFS, such
+  as Infura and Pinata
+- There are no self-describing multiformats for encryption and signing (§3.5.5)
```

Comment:
there are and are communicated as part of the [CID](https://docs.ipfs.io/concepts/content-addressing/#identifier-formats) using [multicodecs](https://github.com/multiformats/multicodec).

Reply:
I looked at those, but couldn't find any encryption or signature schemes in the [multicodecs table](https://github.com/multiformats/multicodec/blob/master/table.csv). I also saw a bunch of open issues in the multiformat repo that lead me to conclude that it's not there yet.

Reply:
`multihash` describes different cryptographic functions, not sure what's missing there? 

>  I also saw a bunch of open issues in the multiformat repo that lead me to conclude that it's not there yet.

Multiformats are used pretty extensively across the ecosystem. For example, libp2p implementation uses multiaddrs exclusively. It is pretty stable. I haven't looked, but a lot of the issues might be related to adding support for new formats?

Reply:
I'd expect to be able to specify that an IPFS object is encrypted with Salsa20 and authenticated with Poly1305 (for example). Having upgradable encryption schemes feels like a natural fit for multiformats, yet I could not find a multiformat for encryption. Multihash for instance only only describes cryptographic hashing functions, not encryption schemes.

Reply:
That would be [multicodec](https://github.com/multiformats/multicodec), it describes the contents "type", with the multicodec [table](https://github.com/multiformats/multicodec/blob/master/table.csv), defining all the supported codecs.

If you read the table it mostly describes application specific codecs, for example ethereum datastructures - https://github.com/multiformats/multicodec/blob/master/table.csv#L54-L62. 

It is possible to store encrypted data of course, but, it would be application specific. This is because there are many ways of doing encryption and hashing/digesting, I suspect that catering to each and every use is a bit too much. For example, where and how to locate the public key/keys, is it included in the blob, or is it stored separately? Also, if the content is encrypted, do you really want to reveal how it was encrypted?

Reply:
Interestingly, draft 3 of the paper describes a native encrypted DAG node type (not sure if it's present in earlier versions). I'll add it here since it's small enough:


--------------------------

### 3.5.5 Object-level Cryptography

IPFS is equipped to handle object-level cryptographic operations. An encrypted or signed object is wrapped in a
special frame that allows encryption or verification of the raw bytes.

```go
type EncryptedObject struct {
  Object []bytes
  // raw object data encrypted
  Tag []bytes
  // optional tag for encryption groups
}

type SignedObject struct {
  Object []bytes
  // raw object data signed
  Signature []bytes
  // hmac signature
  PublicKey []multihash
  // multihash identifying key
}
```

Cryptographic operations change the object’s hash, defining a different object. IPFS automatically verifies signatures, and can decrypt data with user-specified keychains. Links of encrypted objects are protected as well, making traversal impossible without a decryption key. It is possible to have a parent object encrypted under one key, and
a child under another or not at all. This secures links to shared objects.

Reply:
You're right, multicodec is probably the correct way to specify which encryption scheme is used for encrypted content. 

Here's a link to some discussions about signatures and keys in the multiformats repo: [multikey](https://github.com/ipfs/specs/issues/58), [multisig](https://github.com/multiformats/multiformats/issues/23).